src/infiniteLoops/instrument.ts

```diff
@@ -0,0 +1,606 @@
+import * as es from 'estree'
+import { generate } from 'astring'
+import * as create from '../utils/astCreator'
+import { simple, recursive, WalkerCallback } from '../utils/walkers'
+// transforms AST of program
+
+const globalIds = {
+  builtinsId: 'builtins',
+  functionsId: '__InfLoopFns',
+  stateId: '__InfLoopState'
+}
+
+enum FunctionNames {
+  nothingFunction,
+  concretize,
+  hybridize,
+  wrapArg,
+  dummify,
+  saveBool,
+  saveVar,
+  preFunction,
+  returnFunction,
+  postLoop,
+  enterLoop,
+  exitLoop,
+  trackLoc,
+  evalB,
+  evalU
+}
+
+/**
+ * Renames all variables in the program to differentiate shadowed variables and
+ * variables declared with the same name but in different scopes.
+ *
+ * E.g. "function f(f)..." -> "function f_0(f_1)..."
+ * @param predefined A table of [key: string, value:string], where variables named 'key' will be renamed to 'value'
+ */
+function unshadowVariables(program: es.Node, predefined = {}) {
+  for (const name of Object.values(globalIds)) {
+    predefined[name] = name
+  }
+  const seenIds = new Set()
+  const env = [predefined]
+  const genId = (name: string) => {
+    let count = 0
+    while (seenIds.has(`${name}_${count}`)) count++
+    const newName = `${name}_${count}`
+    seenIds.add(newName)
+    env[0][name] = newName
+    return newName
+  }
+  const unshadowFunctionInner = (
+    node: es.FunctionDeclaration | es.ArrowFunctionExpression | es.FunctionExpression,
+    s: undefined,
+    callback: WalkerCallback<undefined>
+  ) => {
+    env.unshift({ ...env[0] })
+    for (const id of node.params as es.Identifier[]) {
+      id.name = genId(id.name)
+    }
+    callback(node.body, undefined)
+    env.shift()
+  }
+  const doStatements = (stmts: es.Statement[], callback: WalkerCallback<undefined>) => {
+    for (const stmt of stmts) {
+      if (stmt.type === 'FunctionDeclaration') {
+        // do hoisting first
+        stmt.id = stmt.id as es.Identifier
+        stmt.id.name = genId(stmt.id.name)
+      } else if (stmt.type === 'VariableDeclaration') {
+        for (const decl of stmt.declarations) {
+          decl.id = decl.id as es.Identifier
+          const newName = genId(decl.id.name)
+          decl.id.name = newName
+        }
+      }
+    }
+    for (const stmt of stmts) {
+      callback(stmt, undefined)
+    }
+  }
+  recursive(program, [{}], {
```

Comment:
Just checking, have you tested renaming with if statements, which can also introduce new envs? My guess is that it's not a problem due to to the BlockStatement introducing a new env, but it's probably prudent to check it in a test anyway.